nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -165,19 +180,19 @@ WaypointFollower::followWaypoints()
 
   while (rclcpp::ok()) {
     // Check if asked to stop processing action
-    if (action_server_->is_cancel_requested()) {
+    if (action_server->is_cancel_requested()) {
       auto cancel_future = nav_to_pose_client_->async_cancel_all_goals();
       rclcpp::spin_until_future_complete(client_node_, cancel_future);
       // for result callback processing
       spin_some(client_node_);
-      action_server_->terminate_all();
+      action_server->terminate_all();
       return;
     }
 
     // Check if asked to process another action
-    if (action_server_->is_preempt_requested()) {
+    if (action_server->is_preempt_requested()) {
       RCLCPP_INFO(get_logger(), "Preempting the goal pose.");
-      goal = action_server_->accept_pending_goal();
+      goal = action_server->accept_pending_goal();
```

Comment:
I'm not sure that how this is written that preemption will be used. You're getting the new `goal` here, but in this `followWaypointsLogic` method, its using the input `poses` variable rather than `goal->poses`, so it wont use the newest poses available when a new goal is accepted. Both `posts` are PoseStamped so that doesn't need to be templated and solves a part of that problem.

The GPS preemption `goal->poses` will be in the wrong coordinate frame type (GPS), so that makes me think that potentially having all this as 1 method might not work, since we need to be able to convert the preempted goal poses into cartesian to set the local variable, the preemption part of this needs "hooks" in to the specific action server being used to run that conversion.

You might be able to get away with it with still 1 `logic`  function if you have one of the inputs a preemption handling lambda (or something) to do that conversion for the GPS case and pass for the non-GPS case. Either way, the `goal` needs to be used / update the `poses` variable.

Reply:
very crucial catch, thumbs up. https://github.com/ros-planning/navigation2/pull/2111/commits/cd9a287063c0f187ef5a544f1d893278f9e67040 tries to deal with that using C++17 specifc future ` if constexpress`. Are we allowed to use C++17 ? , I am aware that ROS2 targets C++14, this can be also be dealth using something C++14 but it looks quite ugly tbh.

Reply:
C++17 can be used when its required, and it looks like you are actually using a C++17 feature so that's acceptable.